README.md
## Hopefully, CCMath is better than CMath...! (・ω・;)

[![Build Status](https://travis-ci.org/gogotanaka/ccmath.svg?branch=master)](https://travis-ci.org/gogotanaka/ccmath)

CCMath is CMath implemented with C-lang.

## Demo
```ruby
CCMath.cos(1i)
#=> (1.5430806348152437+0.0i)

CCMath.cos(1)
#=> (0.5403023058681398+0.0i)
```

## Bench

```
Calculating -------------------------------------
               CMath    28.423k i/100ms
              CCMath    61.123k i/100ms
-------------------------------------------------
               CMath    379.852k (± 4.2%) i/s -      1.904M
              CCMath      1.163M (± 4.0%) i/s -      5.807M

Comparison:
              CCMath:  1163020.7 i/s
               CMath:   379852.3 i/s - 3.06x slower

```

## What's the difference between CCMath and CMath..?(・ω・;)
1. CCMath return Complex object even if image part is equivalent 0.
2. Faster(about 3 times)
3. Better error handling

Other behaviours suppose to be same.

## Installation


```ruby
gem 'ccmath'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install ccmath

## Development

After checking out the repo, run `bin/setup` to install dependencies.

Then, run `bundle exec rake compile` to compile library and `bundle exec rake test` to run the tests.

You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`.

<!-- To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org). -->

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/[USERNAME]/ccmath. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](contributor-covenant.org) code of conduct.


## License

The MIT License (MIT)

Copyright (c) 2015 gogotanaka

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

# Contributor Code of Conduct

As contributors and maintainers of this project, we pledge to respect all people who contribute through reporting issues, posting feature requests, updating documentation, submitting pull requests or patches, and other activities.

We are committed to making participation in this project a harassment-free experience for everyone, regardless of level of experience, gender, gender identity and expression, sexual orientation, disability, personal appearance, body size, race, ethnicity, age, or religion.

Examples of unacceptable behavior by participants include the use of sexual language or imagery, derogatory comments or personal attacks, trolling, public or private harassment, insults, or other unprofessional conduct.

Project maintainers have the right and responsibility to remove, edit, or reject comments, commits, code, wiki edits, issues, and other contributions that are not aligned to this Code of Conduct. Project maintainers who do not follow the Code of Conduct may be removed from the project team.

Instances of abusive, harassing, or otherwise unacceptable behavior may be reported by opening an issue or contacting one or more of the project maintainers.

This Code of Conduct is adapted from the [Contributor Covenant](http://contributor-covenant.org), version 1.0.0, available at [http://contributor-covenant.org/version/1/0/0/](http://contributor-covenant.org/version/1/0/0/)


CHANGELOG.md
## CHANGELOG

### v2.71

* `CCMath.gamma` can receive `Complex` object.
* Now `CCMath.set` is deprecated.

### v2.7

* Initial release.


test/ccmath_test.rb
require 'test_helper'

class CCMathTest < Minitest::Test
  def test_error_handling
    # assert_raises(TypeError, "Numeric Number required") { CMath.acos("2") }
    # assert_raises(TypeError, "Numeric Number required") { CMath.log("2") }
    # assert_raises(ArgumentError) { CMath.log(2, "2") }
    assert_raises(CCMath::DomainError) { CMath.log(2, 2i) }
  end
end


test/test_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'ccmath'

require 'minitest/autorun'


class ::Complex
  def i
    self * 1i
  end
end


test/ccmath_real_test.rb
# require 'test_helper'
#
# class CcmathRealTest < Minitest::Test
#
#   def setup
#     CCMath.set = 'R'
#   end
#
#   def assert_infinity(a, *rest)
#     rest = ["not infinity: #{a.inspect}"] if rest.empty?
#     assert_predicate(a, :infinite?, *rest)
#   end
#
#   def assert_nan(a, *rest)
#     rest = ["not nan: #{a.inspect}"] if rest.empty?
#     assert_predicate(a, :nan?, *rest)
#   end
#
#   def assert_float(a, b)
#     err = [Float::EPSILON * 4, [a.abs, b.abs].max * Float::EPSILON * 256].max
#     assert_in_delta(a, b, err)
#   end
#   alias check assert_float
#
#   def assert_float_and_int(exp_ary, act_ary)
#     flo_exp, int_exp, flo_act, int_act = *exp_ary, *act_ary
#     assert_float(flo_exp, flo_act)
#     assert_equal(int_exp, int_act)
#   end
#
#   def test_atan2
#     check(+0.0, CCMath.atan2(+0.0, +0.0))
#     check(-0.0, CCMath.atan2(-0.0, +0.0))
#     check(+CCMath::PI, CCMath.atan2(+0.0, -0.0))
#     check(-CCMath::PI, CCMath.atan2(-0.0, -0.0))
#
#     inf = Float::INFINITY
#     expected = 3.0 * CCMath::PI / 4.0
#     check(+expected, CCMath.atan2(+inf, -inf))
#     check(-expected, CCMath.atan2(-inf, -inf))
#
#     expected = CCMath::PI / 4.0
#     check(+expected, CCMath.atan2(+inf, +inf))
#     check(-expected, CCMath.atan2(-inf, +inf))
#
#     check(0, CCMath.atan2(0, 1))
#     check(CCMath::PI / 4, CCMath.atan2(1, 1))
#     check(CCMath::PI / 2, CCMath.atan2(1, 0))
#   end
#
#   def test_cos
#     check(1.0,  CCMath.cos(0 * CCMath::PI / 4))
#     check(1.0 / CCMath.sqrt(2), CCMath.cos(1 * CCMath::PI / 4))
#     check(0.0,  CCMath.cos(2 * CCMath::PI / 4))
#     check(-1.0, CCMath.cos(4 * CCMath::PI / 4))
#     check(0.0,  CCMath.cos(6 * CCMath::PI / 4))
#     check(0.5403023058681398,  CCMath.cos(1))
#   end
#
#   def test_sin
#     check(0.0,  CCMath.sin(0 * CCMath::PI / 4))
#     check(1.0 / CCMath.sqrt(2), CCMath.sin(1 * CCMath::PI / 4))
#     check(1.0,  CCMath.sin(2 * CCMath::PI / 4))
#     check(0.0,  CCMath.sin(4 * CCMath::PI / 4))
#     check(-1.0, CCMath.sin(6 * CCMath::PI / 4))
#   end
#
#   def test_tan
#     check(0.0, CCMath.tan(0 * CCMath::PI / 4))
#     check(1.0, CCMath.tan(1 * CCMath::PI / 4))
#     assert_operator(CCMath.tan(2 * CCMath::PI / 4).abs, :>, 1024)
#     check(0.0, CCMath.tan(4 * CCMath::PI / 4))
#     assert_operator(CCMath.tan(6 * CCMath::PI / 4).abs, :>, 1024)
#   end
#
#   def test_acos
#     check(0 * CCMath::PI / 4, CCMath.acos( 1.0))
#     check(1 * CCMath::PI / 4, CCMath.acos( 1.0 / CCMath.sqrt(2)))
#     check(2 * CCMath::PI / 4, CCMath.acos( 0.0))
#     check(4 * CCMath::PI / 4, CCMath.acos(-1.0))
#     assert_raises(CCMath::DomainError) { CCMath.acos(+1.0 + Float::EPSILON) }
#     assert_raises(CCMath::DomainError) { CCMath.acos(-1.0 - Float::EPSILON) }
#     assert_raises(CCMath::DomainError) { CCMath.acos(2.0) }
#   end
#
#   def test_asin
#     check( 0 * CCMath::PI / 4, CCMath.asin( 0.0))
#     check( 1 * CCMath::PI / 4, CCMath.asin( 1.0 / CCMath.sqrt(2)))
#     check( 2 * CCMath::PI / 4, CCMath.asin( 1.0))
#     check(-2 * CCMath::PI / 4, CCMath.asin(-1.0))
#     assert_raises(CCMath::DomainError) { CCMath.asin(+1.0 + Float::EPSILON) }
#     assert_raises(CCMath::DomainError) { CCMath.asin(-1.0 - Float::EPSILON) }
#     assert_raises(CCMath::DomainError) { CCMath.asin(2.0) }
#   end
#
#   def test_atan
#     check( 0 * CCMath::PI / 4, CCMath.atan( 0.0))
#     check( 1 * CCMath::PI / 4, CCMath.atan( 1.0))
#     check( 2 * CCMath::PI / 4, CCMath.atan(1.0 / 0.0))
#     check(-1 * CCMath::PI / 4, CCMath.atan(-1.0))
#   end
#
#   def test_cosh
#     check(1, CCMath.cosh(0))
#     check((CCMath::E ** 1 + CCMath::E ** -1) / 2, CCMath.cosh(1))
#     check((CCMath::E ** 2 + CCMath::E ** -2) / 2, CCMath.cosh(2))
#   end
#
#   def test_sinh
#     check(0, CCMath.sinh(0))
#     check((CCMath::E ** 1 - CCMath::E ** -1) / 2, CCMath.sinh(1))
#     check((CCMath::E ** 2 - CCMath::E ** -2) / 2, CCMath.sinh(2))
#   end
#
#   def test_tanh
#     check(CCMath.sinh(0) / CCMath.cosh(0), CCMath.tanh(0))
#     check(CCMath.sinh(1) / CCMath.cosh(1), CCMath.tanh(1))
#     check(CCMath.sinh(2) / CCMath.cosh(2), CCMath.tanh(2))
#   end
#
#   def test_acosh
#     check(0, CCMath.acosh(1))
#     check(1, CCMath.acosh((CCMath::E ** 1 + CCMath::E ** -1) / 2))
#     check(2, CCMath.acosh((CCMath::E ** 2 + CCMath::E ** -2) / 2))
#     assert_raises(CCMath::DomainError) { CCMath.acosh(1.0 - Float::EPSILON) }
#     assert_raises(CCMath::DomainError) { CCMath.acosh(0) }
#   end
#
#   def test_asinh
#     check(0, CCMath.asinh(0))
#     check(1, CCMath.asinh((CCMath::E ** 1 - CCMath::E ** -1) / 2))
#     check(2, CCMath.asinh((CCMath::E ** 2 - CCMath::E ** -2) / 2))
#   end
#
#   def test_atanh
#     check(0, CCMath.atanh(CCMath.sinh(0) / CCMath.cosh(0)))
#     check(1, CCMath.atanh(CCMath.sinh(1) / CCMath.cosh(1)))
#     check(2, CCMath.atanh(CCMath.sinh(2) / CCMath.cosh(2)))
#     assert_infinity(CCMath.atanh(1))
#     assert_infinity(-CCMath.atanh(-1))
#     assert_raises(CCMath::DomainError) { CCMath.atanh(+1.0 + Float::EPSILON) }
#     assert_raises(CCMath::DomainError) { CCMath.atanh(-1.0 - Float::EPSILON) }
#   end
#
#   def test_exp
#     check(1, CCMath.exp(0))
#     check(CCMath.sqrt(CCMath::E), CCMath.exp(0.5))
#     check(CCMath::E, CCMath.exp(1))
#     check(CCMath::E ** 2, CCMath.exp(2))
#   end
#
#   def test_log
#     check(0, CCMath.log(1))
#     check(1, CCMath.log(CCMath::E))
#     check(0, CCMath.log(1, 10))
#     check(1, CCMath.log(10, 10))
#     check(2, CCMath.log(100, 10))
#     check(CCMath.log(2.0 ** 64), CCMath.log(1 << 64))
#     check(CCMath.log(2) * 1024.0, CCMath.log(2 ** 1024))
#     assert_infinity(CCMath.log(1.0/0))
#     assert_infinity(-CCMath.log(+0.0))
#     assert_infinity(-CCMath.log(-0.0))
#     assert_raises(CCMath::DomainError) { CCMath.log(-1.0) }
#     assert_raises(TypeError) { CCMath.log(1,nil) }
#     assert_raises(CCMath::DomainError, '[ruby-core:62309] [ruby-Bug #9797]') { CCMath.log(1.0, -1.0) }
#     assert_nan(CCMath.log(0.0, 0.0))
#   end
#
#   def test_log2
#     check(0, CCMath.log2(1))
#     check(1, CCMath.log2(2))
#     check(2, CCMath.log2(4))
#     check(CCMath.log2(2.0 ** 64), CCMath.log2(1 << 64))
#     check(1024.0, CCMath.log2(2 ** 1024))
#     assert_infinity(CCMath.log2(1.0/0))
#     assert_infinity(-CCMath.log2(+0.0))
#     assert_infinity(-CCMath.log2(-0.0))
#     assert_raises(CCMath::DomainError) { CCMath.log2(-1.0) }
#   end
#
#   def test_log10
#     check(0, CCMath.log10(1))
#     check(1, CCMath.log10(10))
#     check(2, CCMath.log10(100))
#     check(CCMath.log10(2.0 ** 64), CCMath.log10(1 << 64))
#     check(CCMath.log10(2) * 1024, CCMath.log10(2 ** 1024))
#     assert_infinity(CCMath.log10(1.0/0))
#     assert_infinity(-CCMath.log10(+0.0))
#     assert_infinity(-CCMath.log10(-0.0))
#     assert_raises(CCMath::DomainError) { CCMath.log10(-1.0) }
#   end
#
#   def test_sqrt
#     check(0, CCMath.sqrt(0))
#     check(1, CCMath.sqrt(1))
#     check(2, CCMath.sqrt(4))
#     assert_infinity(CCMath.sqrt(1.0/0))
#     assert_equal("0.0", CCMath.sqrt(-0.0).to_s) # insure it is +0.0, not -0.0
#     assert_raises(CCMath::DomainError) { CCMath.sqrt(-1.0) }
#   end
#
#   def test_cbrt
#     check(1, CCMath.cbrt(1))
#     check(-2, CCMath.cbrt(-8))
#     check(3, CCMath.cbrt(27))
#     check(-0.1, CCMath.cbrt(-0.001))
#     assert_infinity(CCMath.cbrt(1.0/0))
#   end
#
#   def test_frexp
#     assert_float_and_int([0.0,  0], CCMath.frexp(0.0))
#     assert_float_and_int([0.5,  0], CCMath.frexp(0.5))
#     assert_float_and_int([0.5,  1], CCMath.frexp(1.0))
#     assert_float_and_int([0.5,  2], CCMath.frexp(2.0))
#     assert_float_and_int([0.75, 2], CCMath.frexp(3.0))
#   end
#
#   def test_ldexp
#     check(0.0, CCMath.ldexp(0.0, 0.0))
#     check(0.5, CCMath.ldexp(0.5, 0.0))
#     check(1.0, CCMath.ldexp(0.5, 1.0))
#     check(2.0, CCMath.ldexp(0.5, 2.0))
#     check(3.0, CCMath.ldexp(0.75, 2.0))
#   end
#
#   def test_hypot
#     check(5, CCMath.hypot(3, 4))
#   end
#
#   def test_erf
#     check(0, CCMath.erf(0))
#     check(1, CCMath.erf(1.0 / 0.0))
#   end
#
#   def test_erfc
#     check(1, CCMath.erfc(0))
#     check(0, CCMath.erfc(1.0 / 0.0))
#   end
#
#   def test_gamma

#     #
#     # # no SEGV [ruby-core:25257]
#     # 31.upto(65) do |i|
#     #   i = 1 << i
#     #   assert_infinity(CCMath.gamma(i), "CCMath.gamma(#{i}) should be INF")
#     #   assert_infinity(CCMath.gamma(i-1), "CCMath.gamma(#{i-1}) should be INF")
#     # end
#     #
#     # assert_raises(CCMath::DomainError) { CCMath.gamma(-Float::INFINITY) }
#   end
#
#   def test_lgamma
#     # sqrt_pi = CCMath.sqrt(CCMath::PI)
#     # assert_float_and_int([CCMath.log(4 * sqrt_pi / 3),  1], CCMath.lgamma(-1.5))
#     # assert_float_and_int([CCMath.log(2 * sqrt_pi),     -1], CCMath.lgamma(-0.5))
#     # assert_float_and_int([CCMath.log(sqrt_pi),          1], CCMath.lgamma(0.5))
#     # assert_float_and_int([0,                          1], CCMath.lgamma(1))
#     # assert_float_and_int([CCMath.log(sqrt_pi / 2),      1], CCMath.lgamma(1.5))
#     # assert_float_and_int([0,                          1], CCMath.lgamma(2))
#     # assert_float_and_int([CCMath.log(3 * sqrt_pi / 4),  1], CCMath.lgamma(2.5))
#     # assert_float_and_int([CCMath.log(2),                1], CCMath.lgamma(3))
#     # assert_float_and_int([CCMath.log(15 * sqrt_pi / 8), 1], CCMath.lgamma(3.5))
#     # assert_float_and_int([CCMath.log(6),                1], CCMath.lgamma(4))
#     #
#     # assert_raises(CCMath::DomainError) { CCMath.lgamma(-Float::INFINITY) }
#   end
# end


ext/ccmath/ccmath.c
#include "ccmath.h"

VALUE rb_mCCMath;
VALUE rb_eMathDomainError;

static ID id_real_p, id_set, id_power;

#define RB_BIGNUM_TYPE_P(x) RB_TYPE_P((x), T_BIGNUM)
#define BIGNUM_POSITIVE_P(b) (FIX2LONG(rb_big_cmp((b), INT2FIX(0))) >= 0)

#define fix2dbl_without_to_f(x) (double) FIX2LONG(x)
#define big2dbl_without_to_f(x) rb_big2dbl(x)
#define int2dbl_without_to_f(x) \
    (FIXNUM_P(x) ? fix2dbl_without_to_f(x) : big2dbl_without_to_f(x))
#define rat2dbl_without_to_f(x) \
    (int2dbl_without_to_f(rb_rational_num(x)) / int2dbl_without_to_f(rb_rational_den(x)))

static inline double
num2dbl_without_to_f(VALUE num)
{
    if (SPECIAL_CONST_P(num)) {
        if (FIXNUM_P(num)) {
            return fix2dbl_without_to_f(num);
        }
        else if (FLONUM_P(num)) {
            return RFLOAT_VALUE(num);
        }
    }
    else {
        switch (BUILTIN_TYPE(num)) {
          case T_FLOAT:    return RFLOAT_VALUE(num);
          case T_BIGNUM:   return big2dbl_without_to_f(num);
          case T_RATIONAL: return rat2dbl_without_to_f(num);
        }
    }
    return RFLOAT_VALUE(rb_to_float(num));
}

#define NUM2DBL_F(x) num2dbl_without_to_f(x)

inline static VALUE
f_real_p(VALUE x)
{
    return rb_funcall(x, id_real_p, 0);
}

#define binop(n,op) \
inline static VALUE \
f_##n(VALUE x, VALUE y)\
{\
    return rb_funcall(x, (op), 1, y);\
}

binop(add, '+')
binop(div, '/')
binop(sub, '-')
binop(mul, '*')
binop(pow, id_power)

inline static VALUE
DBLS2COMP(double real, double imag)
{
    NEWOBJ_OF(obj, struct RComplex, rb_cComplex,
        T_COMPLEX | (RGENGC_WB_PROTECTED_COMPLEX ? FL_WB_PROTECTED : 0));

    RCOMPLEX_SET_REAL(obj, DBL2NUM(real));
    RCOMPLEX_SET_IMAG(obj, DBL2NUM(imag));

    return (VALUE)obj;
}

#define EXTRACT_DBLS_FROM_REAL(x) \
    x##_real = NUM2DBL_F(x);      \
    x##_imag = 0.0;

#define EXTRACT_DBLS_FROM_COMP(z)        \
    struct RComplex* dat_##z;            \
    dat_##z = ((struct RComplex*)(z));   \
    z##_real = NUM2DBL_F(dat_##z->real); \
    z##_imag = NUM2DBL_F(dat_##z->imag);

#define EXTRACT_DBLS(z);              \
    double z##_real, z##_imag;        \
    do {                              \
        if (f_real_p(z)) {            \
            EXTRACT_DBLS_FROM_REAL(z) \
        } else {                      \
            EXTRACT_DBLS_FROM_COMP(z) \
        }                             \
    } while(0)

#define domain_error(msg) \
    rb_raise(rb_eMathDomainError, "Numerical argument is out of domain - " #msg)

static VALUE
ccmath_sqrt(VALUE obj, VALUE z)
{
    EXTRACT_DBLS(z);
    if (z_imag == 0.0) {
        if (z_real < 0.0) {
            return DBLS2COMP(0.0, sqrt(fabs(z_real)));
        }
        else {
            return DBLS2COMP(sqrt(z_real), 0.0);
        }
    }
    else {
        double s = sqrt((hypot(z_real, z_imag) + z_real) / 2.0);
        return DBLS2COMP(s, z_imag / (2 * s));
    }
}

static VALUE
ccmath_exp(VALUE obj, VALUE z)
{
    EXTRACT_DBLS(z);
    if (z_imag == 0.0) return DBLS2COMP(exp(z_real), 0.0);

    double ere = exp(z_real);
    return DBLS2COMP(ere * cos(z_imag), ere * sin(z_imag));
}

void
assing_numbits(VALUE x, double *p_d, size_t *p_numbits) {
    if (RB_BIGNUM_TYPE_P(x) && BIGNUM_POSITIVE_P(x) && DBL_MAX_EXP <= (*p_numbits = rb_absint_numwords(x, 1, NULL))) {
        *p_numbits -= DBL_MANT_DIG;
        x = rb_big_rshift(x, SIZET2NUM(*p_numbits));
    }
    else {
        *p_numbits = 0;
    }

    *p_d = NUM2DBL_F(x);
}

static double
internal_log(VALUE x)
{
    double d;
    size_t numbits;

    assing_numbits(x, &d, &numbits);

    if (d < 0.0) domain_error("log");
    if (d == 0.0) return -INFINITY;

    return log(d) + numbits * log(2); /* log(d * 2 ** numbits) */
}

static VALUE
ccmath_log(int argc, const VALUE* argv, VALUE obj)
{
    VALUE z, base;
    rb_scan_args(argc, argv, "11", &z, &base);

    EXTRACT_DBLS(z);
    float r = hypot(z_real, z_imag);
    if (argc == 2) {
        if (!rb_respond_to(base, id_real_p)) rb_raise(rb_eTypeError, "Numeric Number required");
        EXTRACT_DBLS(base);
        if (base_imag != 0.0) domain_error("log");
        if (base_real > 0.0) {
            double ln_base = log(base_real);
            return DBLS2COMP(log(r) / ln_base, m_atan2(z_imag, z_real) / ln_base);
        }
        else {
            VALUE ln_base = DBLS2COMP(log(fabs(base_real)), M_PI);
            return f_div(DBLS2COMP(log(r), m_atan2(z_imag, z_real)), ln_base);
        }
    }
    else {
        return DBLS2COMP(log(r), m_atan2(z_imag, z_real));
    }
}

static VALUE
ccmath_log2(VALUE obj, VALUE z)
{
    EXTRACT_DBLS(z);
    float r = hypot(z_real, z_imag);
    return DBLS2COMP(log(r) / M_LN2, m_atan2(z_imag, z_real) / M_LN2);
}

static VALUE
ccmath_log10(VALUE obj, VALUE z)
{
    EXTRACT_DBLS(z);
    float r = hypot(z_real, z_imag);
    return DBLS2COMP(log(r) / M_LN10, m_atan2(z_imag, z_real) / M_LN10);
}

static VALUE
ccmath_cos(VALUE obj, VALUE z)
{
    EXTRACT_DBLS(z);
    if (z_real == 0.0) return DBLS2COMP(cosh(z_imag), 0.0);
    if (z_imag == 0.0) return DBLS2COMP(cos(z_real), 0.0);
    return DBLS2COMP(cos(z_real) * cosh(z_imag), -sin(z_real) * sinh(z_imag));
}

static VALUE
ccmath_sin(VALUE obj, VALUE z)
{
    EXTRACT_DBLS(z);
    if (z_real == 0.0) return DBLS2COMP(0.0, sinh(z_imag));
    if (z_imag == 0.0) return DBLS2COMP(sin(z_real), 0.0);
    return DBLS2COMP(sin(z_real) * cosh(z_imag), cos(z_real) * sinh(z_imag));
}

static VALUE
ccmath_tan(VALUE obj, VALUE z)
{
    return f_div(ccmath_sin(obj, z), ccmath_cos(obj, z));
}

static VALUE
ccmath_cosh(VALUE obj, VALUE z)
{
    EXTRACT_DBLS(z);
    if (z_real == 0.0) return DBLS2COMP(cos(z_imag), 0.0);
    if (z_imag == 0.0) return DBLS2COMP(cosh(z_real), 0.0);
    return DBLS2COMP(cosh(z_real) * cos(z_imag), sinh(z_real) * sin(z_imag));
}

static VALUE
ccmath_sinh(VALUE obj, VALUE z)
{
    EXTRACT_DBLS(z);
    if (z_real == 0.0) return DBLS2COMP(0.0, sin(z_imag));
    if (z_imag == 0.0) return DBLS2COMP(sinh(z_real), 0.0);
    return DBLS2COMP(sinh(z_real) * cos(z_imag), cosh(z_real) * sin(z_imag));
}

static VALUE
ccmath_tanh(VALUE obj, VALUE z)
{
    return f_div(ccmath_sinh(obj, z), ccmath_cosh(obj, z));
}

static VALUE
ccmath_asinh(VALUE obj, VALUE z)
{
    EXTRACT_DBLS(z);

    if (z_imag == 0.0) return DBLS2COMP(asinh(z_real), 0.0);

    VALUE s1 = DBLS2COMP(1.0 + z_imag, -z_real);
    VALUE s2 = DBLS2COMP(1.0 - z_imag, z_real);

    s1 = ccmath_sqrt(obj, s1);
    s2 = ccmath_sqrt(obj, s2);
    EXTRACT_DBLS(s1);
    EXTRACT_DBLS(s2);

    return DBLS2COMP(asinh(s1_real * s2_imag - s2_real * s1_imag), atan2(z_imag, s1_real * s2_real - s1_imag * s2_imag));
}

static VALUE
ccmath_asin(VALUE obj, VALUE z)
{
    EXTRACT_DBLS(z);

    if (z_imag == 0.0) return DBLS2COMP(asin(z_real), 0.0);

    VALUE s = ccmath_asinh(obj, DBLS2COMP(-z_imag, z_real));
    EXTRACT_DBLS(s);
    return DBLS2COMP(s_imag, -s_real);
}

static VALUE
ccmath_acosh(VALUE obj, VALUE z)
{
    EXTRACT_DBLS(z);

    if (z_imag == 0.0) return DBLS2COMP(acosh(z_real), 0.0);

    VALUE s1 = DBLS2COMP(z_real - 1.0, z_imag);
    VALUE s2 = DBLS2COMP(z_real + 1.0, z_imag);

    s1 = ccmath_sqrt(obj, s1);
    s2 = ccmath_sqrt(obj, s2);
    EXTRACT_DBLS(s1);
    EXTRACT_DBLS(s2);

    return DBLS2COMP(asinh(s1_real * s2_real + s1_imag * s2_imag), 2.0 * atan2(s1_imag, s2_real));
}

static VALUE
ccmath_acos(VALUE obj, VALUE z)
{
    EXTRACT_DBLS(z);

    if (z_imag == 0.0) return DBLS2COMP(acos(z_real), 0.0);

    return f_sub(DBL2NUM(M_PI / 2.0), ccmath_asin(obj, z));
}

static VALUE
ccmath_atanh(VALUE obj, VALUE z)
{
    EXTRACT_DBLS(z);

    if (z_imag == 0.0) return DBLS2COMP(atanh(z_real), 0.0);

    double sq_imag = z_imag * z_imag;

    return DBLS2COMP(m_log1p(4.0 * z_real / ((1.0 - z_real) * (1.0 - z_real) + sq_imag)) / 4.0,
                    -m_atan2(-2.0 * z_imag, (1.0 - z_real) * (1.0 + z_real) - sq_imag) / 2.0);
}

static VALUE
ccmath_atan(VALUE obj, VALUE z)
{
    EXTRACT_DBLS(z);

    if (z_imag == 0.0) return DBLS2COMP(atan(z_real), 0.0);

    VALUE s = ccmath_atanh(obj, DBLS2COMP(-z_imag, z_real));
    EXTRACT_DBLS(s);
    return DBLS2COMP(s_imag, -s_real);
}

static VALUE
ccmath_gamma(VALUE obj, VALUE z)
{
    EXTRACT_DBLS(z);
    if (z_real < 0.5) {
        VALUE s1, s2;
        s1 = ccmath_gamma(obj, f_sub(DBL2NUM(1), z));
        s2 = ccmath_sin(obj, f_mul(DBL2NUM(M_PI), z));
        return f_div(DBL2NUM(M_PI), f_mul(s1, s2));
    }
    else {
        static const double lanczos_coef[] = {
            0.99999999999980993,
            676.5203681218851,
            -1259.1392167224028,
            771.32342877765313,
            -176.61502916214059,
            12.507343278686905,
            -0.13857109526572012,
            9.9843695780195716e-6,
            1.5056327351493116e-7
        };
        int i;
        double g, s, x_real, x_imag;

        g = 7.0;
        z_real -= 1;
        x_real = lanczos_coef[0];
        x_imag = 0.0;
        for(i=1; i<g+2; i++) {
            s = lanczos_coef[i] / ((z_real+i) * (z_real+i) + (z_imag * z_imag));
            x_real += s * (z_real+i);
            x_imag -= s * z_imag;
        }
        VALUE x = DBLS2COMP(x_real, x_imag);
        VALUE sqrt_2_pi = DBL2NUM(sqrt(2 * M_PI));
        VALUE t = DBLS2COMP(z_real + g + 0.5, z_imag);
        return f_mul(sqrt_2_pi,
               f_mul(f_pow(t, DBLS2COMP(z_real+0.5, z_imag)),
               f_mul(f_div(DBL2NUM(1), ccmath_exp(obj,t)), x)));

    }
}

// static VALUE
// ccmath_define_set(VALUE obj, VALUE str)
// {
//     rb_ivar_set(obj, id_set, str);
//     return obj;
// }

void Init_ccmath(void)
{
    id_real_p = rb_intern("real?");
    id_power = rb_intern("**");
    id_set = rb_intern("set");
    rb_mCCMath = rb_define_module("CCMath");

    rb_eMathDomainError = rb_define_class_under(rb_mCCMath, "DomainError", rb_eStandardError);
    #ifdef M_PI
        rb_define_const(rb_mCCMath, "PI", DBL2NUM(M_PI));
    #else
        rb_define_const(rb_mCCMath, "PI", DBL2NUM(atan(1.0)*4.0));
    #endif

    #ifdef M_E
        rb_define_const(rb_mCCMath, "E", DBL2NUM(M_E));
    #else
        rb_define_const(rb_mCCMath, "E", DBL2NUM(exp(1.0)));
    #endif

    rb_define_module_function(rb_mCCMath, "sqrt", ccmath_sqrt, 1);
    rb_define_module_function(rb_mCCMath, "exp", ccmath_exp, 1);
    rb_define_module_function(rb_mCCMath, "log", ccmath_log, -1);
    rb_define_module_function(rb_mCCMath, "log2", ccmath_log2, 1);
    rb_define_module_function(rb_mCCMath, "log10", ccmath_log10, 1);
    rb_define_module_function(rb_mCCMath, "cos", ccmath_cos, 1);
    rb_define_module_function(rb_mCCMath, "sin", ccmath_sin, 1);
    rb_define_module_function(rb_mCCMath, "tan", ccmath_tan, 1);
    rb_define_module_function(rb_mCCMath, "cosh", ccmath_cosh, 1);
    rb_define_module_function(rb_mCCMath, "sinh", ccmath_sinh, 1);
    rb_define_module_function(rb_mCCMath, "tanh", ccmath_tanh, 1);
    rb_define_module_function(rb_mCCMath, "asinh", ccmath_asinh, 1);
    rb_define_module_function(rb_mCCMath, "asin", ccmath_asin, 1);
    rb_define_module_function(rb_mCCMath, "acosh", ccmath_acosh, 1);
    rb_define_module_function(rb_mCCMath, "acos", ccmath_acos, 1);
    rb_define_module_function(rb_mCCMath, "atanh", ccmath_atanh, 1);
    rb_define_module_function(rb_mCCMath, "atan", ccmath_atan, 1);
    rb_define_module_function(rb_mCCMath, "gamma", ccmath_gamma, 1);
    // rb_define_module_function(rb_mCCMath, "set=", ccmath_define_set, 1);

    // rb_ivar_set(rb_mCCMath, id_set, rb_str_new2("R"));
}
